### ISSUE TITLE
Abstract class and interface class

### ISSUE THREAD

**Issue by wentinggy** [nus-cs2030/2021-s2#64]
This week's lecture talked about abstract class and interface. However I am still confused as to when to use abstract class when to use interface? Any idea how to differentiate when to use them ?

**Comment by brennanleez-coder** [comment]
Hi wentinggy, for me, I usually plan my codes in writing and diagrams before actually typing my code. I believe this might help you as abstract and Interfaces are all about finding the similarities, abstracting them out, and the putting them in another abstract class or interface. My next step would be to identify if there are similar attributes. Off my head, if there are similar attributes, interfaces would be out of the question as there are not attributes in interfaces and thus abstract would be the way to go. Do note that interfaces, pure interfaces in our module, only allows for abstract methods while abstract classes can have abstract methods, concrete methods and even attributes. Lastly, please be careful as if there is at least one abstract method in the class, the file must be abstract

**Comment by etbf** [comment]
Maybe in certain cases where you can find concrete methods that are the same or very similar across all child classes, its a hint that an abstract class should be used since this concrete method can be written in the abstract class which would not be possible if an interface was used.

**Comment by gableejh** [comment]
I would say that abstract classes should be used when all of the child classes are related to one particular object, i.e. a parent. For instance, a Shape (parent) will have the various different shapes as a child. Whereas for interface, it's more of providing a functionality for classes that may not be related

**Comment by lokehsienjie** [comment]
I agree with brennan about the methodology towards crafting your program design! Thank you for that summary. 
Regarding the question at hand, to add on to gableejh's comment, I would think to imagine abstract classes as a category of an object that doesn't quite exist but have some unifying attributes or properties common to all, such as a shape. In this example, a shape can't exist on its own, it must be defined as a square, triangle etc, but yet has attributes and properties like surface area that can be calculated for all shapes. I imagine an interface, on the other hand, as a rulebook to say that every user that implements it MUST have the defined methods. It doesn't have to be a perceived "parent" of the objects, and neither is there any abstraction of how each method is constructed. For example, java API's Comparable interface or List interface. When considering whether to use interfaces, you may also want to think about whether the user-created object already extends another object. In that case, since it can only have 1 parent, an interface may be more suited for your project. Take for example the WritableDevice interface from PA1 Sample Solutions.

**Comment by wglee99** [comment]
Hi one way that I found quite easy to remember when to use which is that if you need to inherit certain attributes, use Abstract class. If you only have to inherit behaviours, i.e methods, then just use Interface since you can implement more than one interface.

**Comment by Atem729** [comment]
I think abstract classes are used when the classes inheriting it share common properties. While interface are used when you want different classes to be able to use the same methods. In this case, using an interface tells the client that all objects implementing this interface will 100% be able to use the methods in the interface. This means that you can call the same method name of different classes!

**Comment by elizabethh07** [comment]
Use **abstract classes** if :
- there are some related classes that need to share some lines of code then you can put these lines of code within abstract class and this abstract class should be extended by all these related classes.
- can define non-static or non-final fields in abstract class
- the classes that extend an abstract class have many common methods or fields, or require access modifiers other than public (such as protected and private).

Use **interfaces** if :
- there is no common code between 2 classes
- All methods declared within an interface must be implemented by the classes that implements this interface.
- a class can implement more than one interface.
- want to specify the behavior of a particular data type, but not concerned about who implements its behavior.

**Comment by wentinggy** [comment]
Thank you everyone for your help! :-) really appreciate it!
